### DISCUSSION TITLE
How does CeleryKubernetes Executor work?

### DISCUSSION THREAD

**Discussion by amila-desilva** [apache/airflow#27151] [Q&A]
Hi Community!
I am trying to understand how the CeleryKubernetes Executor works using the Airflow Helm Chart. Couple of questions I had:

1. How do I verify Celery Executor is being used to schedule tasks? Under the K8s Pod Spec, I see a pod is created even when a task does not specifically use the K8s Executor (I use queue='kubernetes' for tasks that require K8s Executor).
![image](https://user-images.githubusercontent.com/81698709/196800409-e33b02b5-6f91-400d-b675-973fb83eca65.png)

2. What does the environment variable AIRFLOW_IS_K8S_EXECUTOR_POD do? Currently, it is set to true on all tasks executions (with and without K8s Executor).

Thank you for your time!

**Comment by potiuk** [comment]
1. Should not be. Possibly you have airflow wrongly configured
2. means that this is k8s executor POD

**Reply by amila-desilva** [reply]
Thanks Jarek for your response! Currently, I have only executor: "CeleryKubernetesExecutor" set on the values.yaml file. Do you know what are all the configurations required to set CeleryKubernetes Executor on the values.yaml? Or perhaps any documentation I can refer to?

**Reply by potiuk** [reply]
Just airflow configuration here. https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html and you can run `helms` debugging techniques to dry-run and verify that what you are installing in K8S is what you expect https://helm.sh/docs/chart_template_guide/debugging/ (so for example whether the generated configuration is wht you think is valid). Often there are silly mistakes and typos that lead to that and it's better to use automated tools like that rather than checking it yourself or asking other people to analyse thet. You can also use k9s tool for inspection of your K8S once you get it installed and you can see if the resources, configuration files, secrets, variables are all as you expect them to be.

**Reply by potiuk** [reply]
This is also good exercise to learn how K8s work btw and how airflow maps parameters to actual k8s resources (and frankly that's the only way to see those typos etc. K8S is a leaky abstraction and even if you have an easy start like Helm you need to go deeper and understand the resources etc. if you see any problems. Helm is just a glorified JINJA processor to produce those resources (and manage deploying them) - no more no less, so you will eventually end-up with reading the k8s internals anyway,

But the Helm of ours makes it super easy to have a "working" installation that you can inspect and learn faster about (k9s is a great tool to guide your learnings)

**Reply by amila-desilva** [reply]
Great. Thanks for the recommendations and guidance, Jarek! I looked at my code more closely and it seems that it is working as intended, but you can correct me if I am wrong. 

I have my executor enabled to CeleryKubernetesExecutor in the values.yaml file. In my dag code, I have two Python tasks in one dag: the first using the kubernetes queue and executor_config, and the second is a regular Python task no queue assigned. When the first task runs, it generates a K8s pod spec as specified in the executor_config, and the second seems to have a generic K8s pod spec. My understanding from reading the second K8s pod spec is that is the spec for the airflow worker pod which is also the K8s pod so it makes sense that the AIRFLOW_IS_K8S_EXECUTOR_POD is set to True and there is a label: component: worker which explains that it's the airflow worker. Further, I was able to confirm that the second task gets scheduled on Celery because when I go to the airflow-worker pod logs, the task gets added to the celery dag bag (see image below). What are your thoughts on this? There's limited documentation and examples in relation to using the CeleryK8s Executor, so I mainly found it challenging since there was no base line to compare against.

![image](https://user-images.githubusercontent.com/81698709/199259212-a5587c7a-cf1d-48ed-b55b-d0f3a4fb50b1.png)

**Reply by potiuk** [reply]
Well. Yes. It works as intended. I am not sure if you need more - you can also always track your celery tasks in Celery's flower if you want. This is is the tool to manage the "celery" part. I am not sure what more you would expect from the docs. It's alwasy easy to say that "there is not enought docs" but it's much more difficutt to say "those are the docs that we can add". 

I think you are one of the few people now who have the right experience and (already) knowledge what kind of documentation you would need if you did not know what you know now? And contribute exactly the documentation you need back - to help others? People who have more familiarity with the project are actually one of the worst to write the documentaiton, but the users who already know how things work and just learned it - are the best to do so. And it's super-easy - just go to the page on our docs and click "Suggest a change on this page" - and you will get PR opened, where you will be able to send Pull Request with added documentation - no need for development environment, you can do it straight in the Github UI - it's as easy as opening a discussion or writing such comments as you did.

You can asily become one of the > 2200 contributors to the project. Most of the contributions are like that - people who found something missing, contributed it back. I heartily invite you to do so.

**Reply by amila-desilva** [reply]
Yeah... it was confusing to see a pod spec listed for the celery jobs and not seeing justification for it listed out on the docs. I guess that's also the point of having documentation too is that when new features are built, the docs guide users to use the feature seamlessly and list examples too. This is just some constructive feedback coming from user experience stand-point. But I hear and agree with you 100%. It gets challenging to document every single piece of information. And that's where the community contributions and the discussions help. Thanks for your guidance also!🍻 I'll send in a PR with some docs & examples.

**Reply by potiuk** [reply]
Happy to help with reviewing it :)
